Simplenote/src/main/java/com/automattic/simplenote/Simplenote.java

```diff
@@ -209,31 +224,50 @@ public void onActivityResumed(Activity activity) {
 
                 mIsInBackground = false;
                 AppLog.add(Type.ACTION, "App opened");
+                WorkManager.getInstance(getApplicationContext()).cancelUniqueWork(TAG_SYNC);
+                Log.d("Simplenote.onActivityResumed", "Stopped worker");
             }
         }
 
         @Override
-        public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
+        public void onActivityCreated(@NonNull Activity activity, Bundle savedInstanceState) {
         }
 
         @Override
-        public void onActivityStarted(Activity activity) {
+        public void onActivityStarted(@NonNull Activity activity) {
         }
 
+        @SuppressLint("LongLogTag")
         @Override
-        public void onActivityPaused(Activity activity) {
+        public void onActivityPaused(@NonNull Activity activity) {
+            PeriodicWorkRequest syncWorkRequest = new PeriodicWorkRequest.Builder(
+                SyncWorker.class,
+                PeriodicWorkRequest.MIN_PERIODIC_INTERVAL_MILLIS,
```

Comment:
> we are introducing the syncing logic mostly to make sure the user doesn't lose local data, right? If the first sync succeeds and the user doesn't open the app again - is there even anything to sync?

Syncing is not only local data.  Remote data from other platforms and collaborators can affect the local data and revision history.  Syncing in the background allows the app to update local and remote data to avoid overwriting and revisions out of order.

> I'd like to understand if we really need such an intense sheduling.

Reports we've received vaguely mention leaving the app and coming back to see data is lost.  We haven't received detailed reports to be sure.  We [added application logging](https://github.com/Automattic/simplenote-android/pull/1128) to attempt to troubleshoot the issue ourselves, but that hasn't been released yet.  That's why a short interval was chosen.

> I'm mostly concerned about battery life.

This is the first attempt to avoid data loss and mimic the background syncing, which was [fixed to lower battery usage](https://github.com/Automattic/simplenote-android/pull/414/commits/e88c4eab8e0baa5105535b1babc851591df1a809).  If it proves to increase battery usage, we'll either revert or update these changes like extending the interval.  Ultimately, data loss is more important than battery usage.  Since we aren't getting helpful reports from users and we can't reproduce the issue at all, this will help us understand what the root cause is.

Reply:
>Ultimately, data loss is more important than battery usage.

I can't say I agree with this statement. I'm sure there are solutions which won't drain battery live and won't result in data loss. 

> If it proves to increase battery usage, we'll either revert or update these changes like extending the interval.

Assuming we'll see a huge battery increase. However, usually these things cumulate - one small thing at a time.

>Since we aren't getting helpful reports from users and we can't reproduce the issue at all, this will help us understand what the root cause is.

What if we synced the data once after the user leaves the app and then when they come back to the app?

Reply:
So we already sync our local data and we send them to the server. The only thing which can by synced in this new worker is that we receive data from the server -> do we have any clues which would suggest that this solution will help to "restore/save" the local data?

Reply:
> So we already sync our local data and we send them to the server. The only thing which can by synced in this new worker is that we receive data from the server

That's an assumption we've made in the past, but I don't think it's always the case.  The buckets used to be left started indefinitely, which was thought to be causing high battery consumption even though it was never proven.  Now, we stop the buckets after the app is exited for ten seconds.  During that time, it's not guaranteed sync occurs due to the network connection.

> do we have any clues which would suggest that this solution will help to "restore/save" the local data?

This solution doesn't change how the local data is saved.  It simply starts and stops the buckets to trigger syncing.

Reply:
> > do we have any clues which would suggest that this solution will help to "restore/save" the local data?

> This solution doesn't change how the local data is saved. It simply starts and stops the buckets to trigger syncing.

AFAIU the users can't recover the content even when they go back to the app on the same device right? So AFAICT the issue is that the app doesn't save the data locally before it is exited/killed ==> PersistentStore.save(..) doesn't get executed. Do you think calling "Bucket.start()" will help with that issue? I still feel like we are trying to solve a different issue with the worker. If the issue was that users wouldn't see their data on other devices, I'd understand this solution (I still wouldn't agree with scheduling interval). However, afaik that's not the case, right? Could you please elaborate, why you think it'll help. Thanks!

Reply:
> the issue is that the app doesn't save the data locally before it is exited/killed

This isn't clear from user reports.  We instruct them to check the **_History_** feature since the revisions could be out of the expected order.  Users have reported that they find their "lost" content there.  The problem is that users just say something like "data lost" or "sync broken" and don't contact us.  So, we're not certain if it's lost, out of order, or something else.

The idea behind these changes is to mimic the syncing behavior of the app before we started getting reports of "data lost" or "sync broken" while avoiding the potential high battery usage due to keeping the buckets started indefinitely.  Since we don't have much to go on, changing one thing that supposedly worked previously will help diagnose things.  If it doesn't help, then looking at the Simperium Android library would be the next step.

Reply:
>while avoiding the potential high battery usage 

I don't think we'll be avoiding that - 10s every 50s is a huge battery usage. We'll keeping waking up the device during nights even when it lays still on a table. I personally have to recommend to not merge these changes. We still have ideas which could fix the issue and they are not so drastic.

Also if we don't have any detailed reports - is it a common issue? These changes will affect all users not just those who encountered the issue.

Reply:
> I don't think we'll be avoiding that - 10s every 50s is a huge battery usage.

I think we will be avoiding that.  Just starting/stopping the buckets is not necessarily the battery usage issue.  There's nothing proving what was truly the cause of high battery usage previously, but we believe it's due to a disconnected network, out-of-control buckets trying to sync, or some combination of both.  These changes will avoid both of those situations by only triggering when there is a network connection and starting/stopping the buckets.

> Also if we don't have any detailed reports - is it a common issue?

We don't know.  We have been getting reports on Play Store reviews, but that's not the best gauge of whether it's common or not.  The fact that data loss happens at all is a problem.

Reply:
> I think we will be avoiding that. Just starting/stopping the buckets is not necessarily the battery usage issue

I believe just opening a socket can run down the memory as it wakes the CPU. Of course if you do more, you drain more. It might not be such a huge usage that users would report it. However, I still think apps should try to use as little as possible and the proposed approach unfortunately doesn't follow that. The documentation says: 

```When an app connects to the mobile network in the background, the app wakes up the CPU and turns on the radio. Doing so repeatedly can run down a device's battery.```

I realize WorkManager improves the  battery consumption a lot as it includes several optimization techniques such as batching jobs from different apps. However, afaik it doesn't mean we can freely run a background work every 50s for 10s. 


I just realized one more thing `MIN_PERIODIC_INTERVAL_MILLIS` is set to 15 minutes (at least on Android 9, 10). We are cheating by returning `Result.retry()` from `doWork()`. I believe this is not how the "Result.retry()" was intended to be used. 

```Returns an instance of {@link Result} that can be used to indicate that the work
         * encountered a transient failure and should be retried with backoff specified in
         * {@link WorkRequest.Builder#setBackoffCriteria(BackoffPolicy, long, TimeUnit)}```

So even this is a huge indicator that we are running the task way too often. We didn't encounter any error and we are just misusing this technique so we can run the worker more often than `MIN_PERIODIC_INTERVAL_MILLIS`. I think running the task every 15 minutes is still too often for our use case.